plugins/confirm.cpp

```diff
@@ -501,45 +548,59 @@ struct cls##_hooks : cls::screen_type { \
         INTERPOSE_NEXT(render)(); \
         cls##_instance.render(); \
     } \
-    DEFINE_VMETHOD_INTERPOSE(bool, key_conflict, (df::interface_key key)) \
-    { \
-        return cls##_instance.key_conflict(key) || INTERPOSE_NEXT(key_conflict)(key); \
-    } \
 }; \
 IMPLEMENT_VMETHOD_INTERPOSE_PRIO(cls##_hooks, feed, prio); \
 IMPLEMENT_VMETHOD_INTERPOSE_PRIO(cls##_hooks, render, prio); \
-IMPLEMENT_VMETHOD_INTERPOSE_PRIO(cls##_hooks, key_conflict, prio); \
 static int conf_register_##cls = conf_register(&cls##_instance, {\
     &INTERPOSE_HOOK(cls##_hooks, feed), \
     &INTERPOSE_HOOK(cls##_hooks, render), \
-    &INTERPOSE_HOOK(cls##_hooks, key_conflict), \
 });
 
-#define DEFINE_CONFIRMATION(cls, screen) \
+#define DEFINE_CONFIRMATION(cls, screen, focusString) \
     class confirmation_##cls : public confirmation<df::screen> { \
         virtual string get_id() { static string id = char_replace(#cls, '_', '-'); return id; } \
+        virtual string get_focus_string() { return focusString; } \
+        virtual bool match_prefix() { return focusString[strlen(focusString) - 1] == '*'; } \
     }; \
     IMPLEMENT_CONFIRMATION_HOOKS(confirmation_##cls, 0);
 
 /* This section defines stubs for all confirmation dialogs, with methods
     implemented in plugins/lua/confirm.lua.
     IDs (used in the "confirm enable/disable" command, by Lua, and in the docs)
     are obtained by replacing '_' with '-' in the first argument to DEFINE_CONFIRMATION
+
+    The second argument to DEFINE_CONFIRMATION determines the viewscreen that any
+    intercepted input will be fed to.
+
+    The third argument to DEFINE_CONFIRMATION determines the focus string that will
+    be used to determine if the confirmation should be unpaused. If a confirmation is paused
+    and the focus string is no longer found in the current focus, the confirmation will be
+    unpaused. Focus strings ending in "*" will use prefix matching e.g. "dwarfmode/Info*" would
```

Comment:
Is this true? I thought we were matching prefixes unconditionally. Where is the `*` interpreted?

Reply:
I made `matchFocusString` also take a boolean arg which is `true` by default (defined in the header). If false, it will look for an exact (still case insensitive) match. If "*" is at the end of the focus string in the `DEFINE_CONFIRMATION`, it will set `match_prefix` to true on the confirmation instance, otherwise it'll be false. `match_prefix` is then passed to `matchFocusString` [here](https://github.com/DFHack/dfhack/pull/2721/files#diff-e005088f61392cf1a500dac515066f9cfcb18894fdb97549ead1631aa39668b8R308)